File: apps/dashboard/src/lib/voice-capture.ts
'use client';

// Browser mic capture for voice input: 16 kHz mono PCM16, emitted as it arrives.
//
// getUserMedia + a Web Audio ScriptProcessorNode pull raw Float32 PCM off the
// mic; each block is downsampled to 16 kHz and handed straight to the caller,
// which streams it to /api/asr. A live RMS level drives the waveform.
// ScriptProcessorNode is deprecated but works everywhere including iOS Safari
// and needs no separate worklet module URL.
//
// WARM MIC: the mic stream + AudioContext are kept alive for a short while after a
// recording (WARM_HOLD_MS) and reused by the next one. Opening the mic device
// (getUserMedia) has real latency — enough to clip the first words — so warming it
// means a rapid second recording starts capturing INSTANTLY. Released on idle or
// when the tab is hidden (so the mic indicator doesn't linger indefinitely).
//
// iOS note: startStreaming() MUST be invoked synchronously inside a user gesture
// (the FAB pointerdown / the PTT keydown) so getUserMedia + AudioContext.resume
// are allowed.

import { isNativeShell, setNativeMicActive } from './native-bridge';

const TARGET_RATE = 16_000;
// Keep the mic warm briefly after a recording so a back-to-back recording starts
// instantly (no getUserMedia device-open latency → no clipped first words). This is
// only for the clip-fix — NOT a workaround for iOS's per-getUserMedia permission
// re-prompt. That one is unfixable from here: in Safari and in an installed PWA the
// prompt returns on every call, and the warm window merely hides it for 20 s. The
// fix lives in the native shell (apps/ios), where the host app answers the capture
// request itself so the web layer stops being asked at all.
const WARM_HOLD_MS = 20_000;

// ── Warm mic (module-level, shared across recordings) ───────────────────────
let warm: { stream: MediaStream; ctx: AudioContext } | null = null;
let warmTimer: ReturnType<typeof setTimeout> | null = null;
let visibilityBound = false;

// ── Mic permission state ────────────────────────────────────────────────────
// "Would opening the mic RIGHT NOW pop a permission prompt?" — the button needs
// that answer SYNCHRONOUSLY at pointerdown, before it knows whether the touch is
// a press-to-talk or a drag, so the async answer is cached here.
//
// It genuinely changes on its own: iOS never persists a getUserMedia grant. WebKit
// keeps grants in memory (UserMediaPermissionRequestManagerProxy::m_grantedRequests)
// and a watchdog drops them ~10 min after capture stops (24 h while a track is
// live); a reload or a cold app launch starts with none. The Permissions API is
// wired to exactly that state — shouldChangePromptToGrantForMicrophone() answers
// 'granted' iff a live grant matches — so it is the right thing to ask, but the
// answer expires, hence the freshness window below.
export type MicPermission = 'granted' | 'prompt' | 'denied' | 'unknown';
const PERM_FRESH_MS = 20_000;
let micPerm: MicPermission = 'unknown';
let micPermAt = 0;
let permSubscribed = false;

/** True while a mic track from a previous recording is still open (grant is current). */
export function isMicWarm(): boolean {
  return (
    !!warm &&
    warm.ctx.state !== 'closed' &&
    warm.stream.getTracks().some((t) => t.readyState === 'live')
  );
}

/**
 * Last known permission state, synchronously. 'unknown' when the browser can't
 * answer (Permissions API without the 'microphone' name) or the cached answer has
 * gone stale — callers should treat 'unknown' as "can't tell", not as a denial.
 */
export function micPermission(): MicPermission {
  if (isMicWarm()) return 'granted';
  if (micPerm !== 'unknown' && Date.now() - micPermAt > PERM_FRESH_MS) return 'unknown';
  return micPerm;
}

/** Re-ask the browser; safe to call often (cheap, and it drives micPermission()). */
export async function refreshMicPermission(): Promise<MicPermission> {
  try {
    const status = await navigator.permissions?.query({ name: 'microphone' as PermissionName });
    if (!status) return micPerm;
    micPerm = status.state as MicPermission;
    micPermAt = Date.now();
    if (!permSubscribed) {
      permSubscribed = true;
      status.addEventListener('change', () => {
        micPerm = status.state as MicPermission;
        micPermAt = Date.now();
      });
    }
  } catch {
    // Safari before the Permissions API knew 'microphone', Firefox, etc.
  }
  return micPerm;
}

function noteGranted() {
  micPerm = 'granted';
  micPermAt = Date.now();
  lastCaptureAt = Date.now();
  // A stream is open. In the native shell this is what puts iOS's audio session
  // into a category that can actually record.
  setNativeMicActive(true);
}

// How long after a successful capture we still assume the grant is live, when the
// browser won't tell us. Deliberately under WebKit's ~10 min post-capture watchdog.
const CAPTURE_EVIDENCE_MS = 4 * 60_000;
let lastCaptureAt = 0;

/**
 * "Can we open the mic right now WITHOUT a permission prompt?" — the question the
 * button asks on pointerdown, when it still can't tell a press-to-talk from a drag.
 * False means: don't open anything yet, offer an explicit tap-to-allow instead.
 */
export function canOpenMicSilently(): boolean {
  // Inside the native shell the host app answers the capture request itself
  // (WKUIDelegate → .grant), so there is no prompt to protect the user from —
  // and this is the app whose entire reason for existing is not having one.
  // Without this, press-and-hold in the shell demands a redundant "tap to allow"
  // on every cold launch and after every ~10 minutes idle.
  if (isNativeShell()) return true;
  if (isMicWarm()) return true;
  const perm = micPermission();
  if (perm === 'granted') return true;
  if (perm === 'prompt' || perm === 'denied') return false;
  // Browser can't answer (old Safari, Firefox) or the cached answer went stale —
  // fall back to evidence: a capture this recently means the grant is still live.
  return Date.now() - lastCaptureAt < CAPTURE_EVIDENCE_MS;
}

/**
 * Ask for mic access WITHOUT starting a recording — the first-run "tap to allow"
 * step. Must be called straight from a user gesture (a click/pointerup handler):
 * WebKit only treats a request made inside the gesture's own call stack as
 * privileged, and an unprivileged one that was denied before is auto-denied.
 * Leaves the mic warm so the press that follows starts instantly.
 */
export async function requestMicAccess(): Promise<void> {
  await acquireWarm();
  scheduleWarmRelease();
}

function releaseWarm() {
  if (warmTimer) { clearTimeout(warmTimer); warmTimer = null; }
  if (warm) {
    warm.stream.getTracks().forEach((t) => t.stop());
    void warm.ctx.close();
    warm = null;
    // Hand the audio route back — otherwise the shell keeps other apps ducked.
    setNativeMicActive(false);
  }
}

function scheduleWarmRelease() {
  if (warmTimer) clearTimeout(warmTimer);
  warmTimer = setTimeout(releaseWarm, WARM_HOLD_MS);
}

/** Release the warm mic now (call when leaving the chat so it doesn't linger). */
export function releaseWarmMic(): void {
  releaseWarm();
}

async function acquireWarm(): Promise<{ stream: MediaStream; ctx: AudioContext }> {
  if (warmTimer) { clearTimeout(warmTimer); warmTimer = null; } // recording again — cancel pending release
  if (warm && warm.ctx.state !== 'closed' && warm.stream.getTracks().some((t) => t.readyState === 'live')) {
    return warm;
  }
  releaseWarm(); // stale (device revoked / ctx closed) — start fresh

  let stream: MediaStream;
  try {
    stream = await navigator.mediaDevices.getUserMedia({
      audio: { channelCount: 1, echoCancellation: true, noiseSuppression: true },
    });
  } catch (e) {
    // NotAllowedError is the user (or a policy) saying no — remember it so the
    // button offers the "go enable it in Settings" hint instead of re-prompting.
    if ((e as DOMException)?.name === 'NotAllowedError') {
      micPerm = 'denied';
      micPermAt = Date.now();
    }
    throw e;
  }
  noteGranted();
  const Ctx =
    window.AudioContext ||
    (window as unknown as { webkitAudioContext: typeof AudioContext }).webkitAudioContext;
  const ctx = new Ctx();
  if (ctx.state === 'suspended') await ctx.resume();
  warm = { stream, ctx };

  if (!visibilityBound) {
    visibilityBound = true;
    document.addEventListener('visibilitychange', () => { if (document.hidden) releaseWarm(); });
    window.addEventListener('pagehide', releaseWarm);
  }
  return warm;
}

// ── Streaming capture (realtime dictation) ──────────────────────────────────
//
// The batch recorder above hoards Float32 until stop(); this one emits 16 kHz
// PCM16 as it goes, ~85 ms per block (one ScriptProcessor buffer), which is what
// the /api/asr socket forwards to DashScope. Same warm-mic machinery, same
// gesture rules — only the sink differs.
//
// TWO THINGS THIS DOES BEYOND FORWARDING:
//
// · A SILENCE GATE. DashScope bills by the audio second, and a dictation bar left
//   open while nobody talks would stream billable nothing. So after SILENCE_TAIL_MS
//   below the speech threshold we stop emitting until sound returns. The tail is
//   deliberately longer than the server's 800 ms sentence-close silence, so the
//   sentence in flight always gets the quiet it needs to close BEFORE the gate
//   shuts — otherwise the last sentence of every paragraph would hang open.
//
// · A FALLBACK BUFFER. Everything emitted since the last mark() is also kept
//   locally, so if the socket dies mid-dictation the words that were in the air
//   are not lost: the widget POSTs them to the batch /api/transcribe instead.
//   mark() is called on every closed sentence, so this is normally a few seconds
//   of audio, not the whole run.

/** RMS below this is "not speech". Same scale as the level callback ÷ 5. */
const SILENCE_RMS = 0.012;
/** Quiet for this long → stop emitting. Must exceed the server's max_sentence_silence. */
const SILENCE_TAIL_MS = 1_500;

export interface VoiceStream {
  /** Audio emitted since the last mark(), as a WAV — the socket-died fallback clip. */
  stop(): Promise<Blob>;
  /** "Everything so far is safely transcribed" — drops it from the fallback clip. */
  mark(): void;
  /** Abort; no Blob, mic released. */
  cancel(): void;
}

interface StreamOpts {
  /** 16 kHz mono PCM16, ~85 ms per call. Not called while the silence gate is shut. */
  onChunk: (pcm: Int16Array) => void;
  onLevel?: (level: number) => void;
  /** Gate transitions — true when we stopped emitting because nobody is talking. */
  onSilence?: (silent: boolean) => void;
  maxMs?: number;
  onAutoStop?: () => void;
}

/**
 * Stateful decimator. The batch path can resample the whole recording at once;
 * a stream cannot, because 4096 input samples is not a whole number of output
 * samples (48 kHz → 16 kHz leaves a remainder every block). The unconsumed tail
 * carries into the next block, so no click is introduced at the seams.
 */
function makeDownsampler(from: number, to: number): (block: Float32Array) => Float32Array {
  if (from === to) return (b) => b;
  const ratio = from / to;
  let carry = new Float32Array(0);
  return (block: Float32Array) => {
    let buf: Float32Array;
    if (carry.length) {
      buf = new Float32Array(carry.length + block.length);
      buf.set(carry, 0);
      buf.set(block, carry.length);
    } else {
      buf = block;
    }
    const outLen = Math.floor(buf.length / ratio);
    const out = new Float32Array(outLen);
    let n = 0;
    for (let i = 0; i < outLen; i++) {
      const start = Math.floor(i * ratio);
      const end = Math.floor((i + 1) * ratio);
      if (end > buf.length) break;
      let sum = 0;
      for (let j = start; j < end; j++) sum += buf[j];
      out[n++] = sum / (end - start);
    }
    const consumed = Math.floor(n * ratio);
    carry = buf.slice(consumed);
    return n === outLen ? out : out.subarray(0, n);
  };
}

function toPcm16(f32: Float32Array): Int16Array {
  const out = new Int16Array(f32.length);
  for (let i = 0; i < f32.length; i++) {
    const s = Math.max(-1, Math.min(1, f32[i]));
    out[i] = s < 0 ? s * 0x8000 : s * 0x7fff;
  }
  return out;
}

export async function startStreaming(opts: StreamOpts): Promise<VoiceStream> {
  const { stream, ctx } = await acquireWarm();
  if (ctx.state === 'suspended') await ctx.resume();

  const source = ctx.createMediaStreamSource(stream);
  const processor = ctx.createScriptProcessor(4096, 1, 1);
  const mute = ctx.createGain();
  mute.gain.value = 0;

  const down = makeDownsampler(ctx.sampleRate, TARGET_RATE);
  let fallback: Float32Array[] = []; // 16 kHz mono, since the last mark()
  let stopped = false;
  let lastLoudAt = Date.now();
  let gated = false;

  const maxMs = opts.maxMs ?? 30 * 60_000;
  const autoTimer = setTimeout(() => { if (!stopped) opts.onAutoStop?.(); }, maxMs);

  processor.onaudioprocess = (e) => {
    const input = e.inputBuffer.getChannelData(0);
    let sum = 0;
    for (let i = 0; i < input.length; i++) sum += input[i] * input[i];
    const rms = Math.sqrt(sum / input.length);
    opts.onLevel?.(Math.min(1, rms * 5));

    const now = Date.now();
    if (rms >= SILENCE_RMS) lastLoudAt = now;
    const shouldGate = now - lastLoudAt > SILENCE_TAIL_MS;
    if (shouldGate !== gated) {
      gated = shouldGate;
      opts.onSilence?.(gated);
    }
    if (gated) return;

    const pcm = down(new Float32Array(input));
    if (!pcm.length) return;
    fallback.push(pcm.slice());
    opts.onChunk(toPcm16(pcm));
  };

  source.connect(processor);
  processor.connect(mute);
  mute.connect(ctx.destination);

  const teardown = (keepWarm: boolean) => {
    stopped = true;
    clearTimeout(autoTimer);
    processor.onaudioprocess = null;
    try {
      processor.disconnect();
      mute.disconnect();
      source.disconnect();
    } catch {
      /* already gone */
    }
    if (keepWarm) scheduleWarmRelease();
    else releaseWarm();
  };

  return {
    async stop() {
      if (!stopped) teardown(true);
      const total = fallback.reduce((n, c) => n + c.length, 0);
      const merged = new Float32Array(total);
      let off = 0;
      for (const c of fallback) { merged.set(c, off); off += c.length; }
      return encodeWav(merged, TARGET_RATE);
    },
    mark() {
      fallback = [];
    },
    cancel() {
      if (!stopped) teardown(false);
      fallback = [];
    },
  };
}

// Encode mono Float32 PCM as a 16-bit little-endian PCM WAV Blob.
function encodeWav(pcm: Float32Array, rate: number): Blob {
  const buffer = new ArrayBuffer(44 + pcm.length * 2);
  const view = new DataView(buffer);
  const writeStr = (o: number, s: string) => {
    for (let i = 0; i < s.length; i++) view.setUint8(o + i, s.charCodeAt(i));
  };
  writeStr(0, 'RIFF');
  view.setUint32(4, 36 + pcm.length * 2, true);
  writeStr(8, 'WAVE');
  writeStr(12, 'fmt ');
  view.setUint32(16, 16, true); // fmt chunk size
  view.setUint16(20, 1, true); // PCM
  view.setUint16(22, 1, true); // mono
  view.setUint32(24, rate, true);
  view.setUint32(28, rate * 2, true); // byte rate = rate * blockAlign
  view.setUint16(32, 2, true); // block align = channels * bytes/sample
  view.setUint16(34, 16, true); // bits per sample
  writeStr(36, 'data');
  view.setUint32(40, pcm.length * 2, true);
  let off = 44;
  for (let i = 0; i < pcm.length; i++) {
    const s = Math.max(-1, Math.min(1, pcm[i]));
    view.setInt16(off, s < 0 ? s * 0x8000 : s * 0x7fff, true);
    off += 2;
  }
  return new Blob([buffer], { type: 'audio/wav' });
}

File: apps/ios/Hermit/VoiceCapture.swift
import AVFoundation
import Foundation

/// The microphone, as 16 kHz mono PCM16 blocks and a loudness number.
///
/// The native half of `apps/dashboard/src/lib/voice-capture.ts`, and the reason
/// this shell exists at all: WebKit on iOS never persists a `getUserMedia`
/// grant, so the web layer is re-prompted on every cold launch and again about
/// ten minutes after each capture. Here the app owns the grant, iOS remembers
/// it, and the prompt happens once.
///
/// ## What is deliberately not copied
///
/// The web keeps the mic warm for 20 seconds after a run (`WARM_HOLD_MS`) to
/// hide `getUserMedia`'s device-open latency, which clips the first syllables.
/// That latency is a browser-permission round trip; `AVAudioEngine.start()` on
/// an already-configured session is milliseconds, so the warm window would be a
/// live microphone indicator kept on for nothing. The audio SESSION is left
/// configured between runs, which is the part that actually costs time.
///
/// ## Why a converter and not a downsample by hand
///
/// The hardware format is whatever the route says — 48 kHz on the phone, 44.1
/// through some Bluetooth headsets, and it CHANGES when a headset is plugged in
/// mid-run. `AVAudioConverter` is the only thing that resamples correctly for
/// all of them; picking every third sample is aliasing, and ASR hears aliasing
/// as consonants that were never said.
final class VoiceCapture {
    /// What `/api/asr` wants, and what `dashscope` is configured for.
    static let sampleRate: Double = 16_000

    /// One block of 16 kHz mono PCM16, little-endian — ready for the socket.
    private let onChunk: (Data) -> Void
    /// RMS of the block, 0…1, for the blob and the dot.
    private let onLevel: (Double) -> Void

    private let engine = AVAudioEngine()
    private var converter: AVAudioConverter?
    private var sourceFormat: AVAudioFormat?
    private var running = false

    /// The format everything downstream is written against.
    private static let target = AVAudioFormat(commonFormat: .pcmFormatInt16,
                                              sampleRate: VoiceCapture.sampleRate,
                                              channels: 1,
                                              interleaved: true)!

    init(onChunk: @escaping (Data) -> Void, onLevel: @escaping (Double) -> Void) {
        self.onChunk = onChunk
        self.onLevel = onLevel
    }

    // MARK: - Permission

    /// Would opening the mic right now put a system alert on screen?
    ///
    /// The web's `canOpenMicSilently()`, and the answer the press-and-hold needs
    /// SYNCHRONOUSLY at touch-down — before it knows whether the finger is
    /// holding or scrolling. Unlike the browser's, this answer does not expire.
    static var authorized: Bool {
        AVAudioApplication.shared.recordPermission == .granted
    }

    static var denied: Bool {
        AVAudioApplication.shared.recordPermission == .denied
    }

    /// Ask. Safe to call when already granted (it answers immediately).
    ///
    /// Fired from the RELEASE of a hold, never from under a held finger: an
    /// alert raised there swallows the touch and the release that would have
    /// ended the run never arrives. That is the web's rule too, for the same
    /// reason and a different alert.
    static func requestAccess() async -> Bool {
        await AVAudioApplication.requestRecordPermission()
    }

    // MARK: - Running

    /// Open the mic and start delivering blocks.
    ///
    /// Throws rather than failing quietly: every caller has something to say to
    /// the user, and a run that silently records nothing is the failure this
    /// whole feature is least able to survive.
    func start() throws {
        guard !running else { return }

        let session = AVAudioSession.sharedInstance()
        // `.playAndRecord` and not `.record`: a turn's reply can start speaking
        // (Live Activity sounds, and the page's own audio) while dictation is
        // open, and `.record` silences all of it. `.duckOthers` is what the
        // system does to a phone call's other audio, which is the closest thing
        // to what is happening here.
        try session.setCategory(.playAndRecord,
                                mode: .spokenAudio,
                                options: [.duckOthers, .defaultToSpeaker, .allowBluetoothHFP])
        try session.setActive(true, options: [])

        let input = engine.inputNode
        let format = input.inputFormat(forBus: 0)
        // A zero sample rate is what the input node reports when the route is not
        // ready yet — starting a converter against it crashes inside CoreAudio
        // rather than throwing.
        guard format.sampleRate > 0 else { throw VoiceCaptureError.noInput }
        sourceFormat = format
        converter = AVAudioConverter(from: format, to: Self.target)
        guard converter != nil else { throw VoiceCaptureError.noConverter }

        // 100 ms of audio per block at the SOURCE rate. Small enough that the
        // first partial arrives while you are still speaking, large enough that
        // the socket is not writing sixty times a second.
        let frames = AVAudioFrameCount(format.sampleRate / 10)
        input.installTap(onBus: 0, bufferSize: frames, format: format) { [weak self] buffer, _ in
            self?.handle(buffer)
        }
        engine.prepare()
        try engine.start()
        running = true
    }

    /// Close the mic. Idempotent, and safe from any thread.
    func stop() {
        guard running else { return }
        running = false
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        converter = nil
        // Hand the route back so other audio comes out of its duck. Not fatal if
        // it fails — the tap is already gone, so nothing is still listening.
        try? AVAudioSession.sharedInstance().setActive(false, options: [.notifyOthersOnDeactivation])
    }

    var isRunning: Bool { running }

    // MARK: - The tap

    private func handle(_ buffer: AVAudioPCMBuffer) {
        guard let converter, let sourceFormat else { return }
        onLevel(Self.rms(buffer))

        // Capacity for the resampled block, rounded up: 48000 → 16000 is exact,
        // 44100 → 16000 is not, and one frame short throws away a millisecond of
        // every block for the length of the run.
        let ratio = Self.sampleRate / sourceFormat.sampleRate
        let capacity = AVAudioFrameCount((Double(buffer.frameLength) * ratio).rounded(.up)) + 1
        guard let out = AVAudioPCMBuffer(pcmFormat: Self.target, frameCapacity: capacity) else { return }

        var handed = false
        var error: NSError?
        converter.convert(to: out, error: &error) { _, status in
            // Exactly one input buffer per output block; saying `.haveData`
            // twice would hand the same audio over twice.
            if handed { status.pointee = .noDataNow; return nil }
            handed = true
            status.pointee = .haveData
            return buffer
        }
        guard error == nil, out.frameLength > 0,
              let channel = out.int16ChannelData else { return }
        let bytes = Int(out.frameLength) * MemoryLayout<Int16>.size
        onChunk(Data(bytes: channel[0], count: bytes))
    }

    /// Root mean square, scaled so ordinary speech lands around 0.3–0.7.
    ///
    /// The same shape as the web's `publishMicLevel`: what this drives is a blob
    /// that breathes, so it wants to look like loudness rather than be a
    /// measurement. Read off the SOURCE buffer, before conversion, so a
    /// converter that fails still leaves the blob moving.
    private static func rms(_ buffer: AVAudioPCMBuffer) -> Double {
        guard let data = buffer.floatChannelData, buffer.frameLength > 0 else { return 0 }
        let n = Int(buffer.frameLength)
        var sum: Double = 0
        for i in 0..<n {
            let v = Double(data[0][i])
            sum += v * v
        }
        let level = (sum / Double(n)).squareRoot()
        // × 6 and clamped: raw RMS on a phone mic sits under 0.1 for normal
        // speech, and a blob that never grows says the mic is dead.
        return min(1, level * 6)
    }
}

enum VoiceCaptureError: LocalizedError {
    case noInput
    case noConverter

    var errorDescription: String? {
        switch self {
        case .noInput: return "麦克风不可用"
        case .noConverter: return "麦克风格式不支持"
        }
    }
}
